### ISSUE TITLE
Inefficient array clearing of DefaultArrayPool.Return()

### ISSUE THREAD

**Issue by Kaktusbot** [dotnet/corefx#29848]
I was benchmarking things and found out that DefaultArrayPool clearing of array is extremely inefficient compared to simple ```new Span(array).Clear()```
https://github.com/dotnet/corefx/blob/8f968b9e79c5721d78cc88e46fe57457fe9d490d/src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs#L142-L145

My benchmark code
```C#
const int MinSize = 10000;
[Benchmark] 
public void ArrayPoolClearDefault() {
    var b = ArrayPool<byte>.Shared.Rent(MinSize);
    ArrayPool<byte>.Shared.Return(b, true);
}
[Benchmark] 
public void ArrayPoolClearFill0() {
    var b = ArrayPool<byte>.Shared.Rent(MinSize);
    new Span<byte>(b).Fill(0);
    ArrayPool<byte>.Shared.Return(b, false);
}
[Benchmark] 
public void ArrayPoolClearClear() {
    var b = ArrayPool<byte>.Shared.Rent(MinSize);
    new Span<byte>(b).Clear();
    ArrayPool<byte>.Shared.Return(b, false);
}
```
Here is result
```
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i5-3230M CPU 2.60GHz (Ivy Bridge), 1 CPU, 4 logical and 2 physical cores
Frequency=2533310 Hz, Resolution=394.7405 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008673
  [Host]     : .NET Core 2.0.7 (CoreCLR 4.6.26328.01, CoreFX 4.6.26403.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.7 (CoreCLR 4.6.26328.01, CoreFX 4.6.26403.03), 64bit RyuJIT


                Method |       Mean |      Error |    StdDev |
---------------------- |-----------:|-----------:|----------:|
 ArrayPoolClearDefault | 4,241.5 ns | 105.129 ns | 98.337 ns |
   ArrayPoolClearFill0 |   611.4 ns |  14.166 ns | 23.669 ns |
   ArrayPoolClearClear |   590.8 ns |   4.181 ns |  3.911 ns |
```

I don't really understand about how GC works and maybe Array.Clear needed to somehow notify GC that by-ref elements of array are unused anymore.
But if array consists of structs without references then clearing of it should be speeded up.

**Comment by stephentoub** [comment]
What results do you get if you use `object` instead of `byte`?

**Comment by stephentoub** [comment]
@ahsonkhan, @jkotas, https://github.com/dotnet/coreclr/pull/9598 optimized Span.Clear.  Can similar optimizations be applied to Array.Clear, at least in some situations?

**Comment by kasper3** [comment]
> `Span<object>.Fill(0)` was boxing `int = 0`

could that be due to `nuint length = (uint)_length` cast in https://github.com/dotnet/coreclr/blob/32ab54d345a4ab2eb572fe3ba061b6e0d1b31088/src/System.Private.CoreLib/shared/System/Span.Fast.cs#L195-L231?

**Comment by omariom** [comment]
Adding typed overloads of Array.Clear would cure most of the  call sites without much effort.

**Comment by jkotas** [comment]
> Can similar optimizations be applied to Array.Clear, at least in some situations?

https://github.com/dotnet/coreclr/pull/18101

**Comment by danmosemsft** [comment]
@jkotas I belivee this is all fixed, and ported to 2.1. I see you reopened? Please do so again if that was intentional and there is more to do here.
